net: add full CoRE Resource Directory endpoint implementation

haukepetersen opened the pull request:
Waiting on and rebased on a couple of PRs, e.g. ~~#7402~~, ~~#7406~~, ~~#7407~~, ~~#7408~~

This PR adds complete CoRE Resource Directory (RD) functionality for RIOT. The code should be fully functional and is basically done, though it needs some more documentation and some cleanups (and proper squashing of course). Will take that on once all the dependencies are merged...

PeterKietzmann commented on examples/coap_rdcli/README.md:
````diff
@@ -0,0 +1,20 @@
+redefine `RDCLI_SERVER_ADDR` and `RDCLI_SERVER_PORT`, e.g.:
+```
+CFLAGS="-DRDCLI_SERVER_ADDR=IPV6_ADDR_ALL_ROUTERS_LINK_LOCAL" make all
+```
````
Currently this README equals that of the other application. Would be great to have a short introduction what the difference is.

haukepetersen replied:
This is because I have not written an readme for this example, yet :-)

PeterKietzmann commented on sys/shell/commands/sc_rdcli.c:
```diff
@@ -0,0 +1,108 @@
+        rdcli_dump_status();
+    }
+    else {
+        printf("usage: %s <reg [addr [port]]|up|rem|info>\n", argv[0]);
```
I would prefer `update` as command. `up` reminds me of initiating something.

haukepetersen replied:
will do

PeterKietzmann commented on sys/net/application_layer/rdcli/rdcli.c:
```diff
@@ -0,0 +1,271 @@
+    sock_udp_ep_t remote;
+} rd_entry_t;
+
+static rd_entry_t active_rd = { 0 };
```
I'm using a rather old compiler.  arm-none-eabi-gcc 4.9.3 complains `error: missing braces around initializer`

PeterKietzmann commented on sys/net/application_layer/rdcli/rdcli.c:
```diff
@@ -0,0 +1,271 @@
+    /* build CoAP request packet */
+    int res = gcoap_req_init(&pkt, buf, sizeof(buf), COAP_METHOD_POST, "/rd");
+    if (res < 0) {
+        goto rdcli_register;
```
Didn't we decide against "goto" some time ago?

haukepetersen replied:
AFAIK not in general, and though I am still not a huge fan of this, I think for certain error handling cases it can make the code more readable, if used with caution...

haukepetersen commented:
> With examples/coap_rdcli it appeared that ping6 raises a failed assertion

Yepp, thats the missing message queue (once more)... Will fix.

> Could you add a description about how to set up an RD server for easier testing of the actual protocol?

quick version: compile nanocoap and run the `coap-rd` command, then watch with wireshark. But I have not really played with the lookup API (if it is even implenented in libcoap) to see if entries are made correctly.

Alternative: You could also use the javscript RD Pekka wrote in Dagstuhl (https://github.com/Ell-i/coap-rd):
- install node 8.x on your host
- run it using `node coap-rd`
- watch the debug output...

haukepetersen commented:
finally, no more dependencies to wait for. Heavily rebased, and partly reworked this PR, now it is ready for (final) review and testing. 

I tested this mainly by registering it with the RD provided by `aiocoap` (https://github.com/chrysn/aiocoap). For this I always used the latest master of `aiocoap-rd`, as the latest release does not include many features of the latest draft version...

haukepetersen commented:
rebased and fixed some issues that Murdock pointed out.

haukepetersen commented:
@bergzand @PeterKietzmann anyone available to test and review this PR?!

PeterKietzmann commented:
Unfortunately I won't find time for this soon. @pokgak are you familiar with the idea of "Resource Directory"? Isn't this somewhat interesting for you?

pokgak commented:
Perfect timing. Currently reading on this topic to use it in my project. Will test it next week and provide feedback.

haukepetersen commented:
@pokgak awesome to hear, looking forward to your feedback!

bergzand commented:
@haukepetersen I hope to have some time for a review next week somewhere.

pokgak commented:
Tested this using native with PORT=tap0, aiocoap-rd bind to global address of tapbr0 and Wireshark listens on tap0. 

Here is what I've tested:
- Running discover, reg, update, info, rem directly after each other work as expected OK
- `rdcli update` and `rdcli rem` when not registered does not do anything. OK
- Client updates automatically once before the lifetime is over, but not anymore after that.
  Steps to reproduce:
  + `rdcli reg [rd_ip]`, then wait and see wireshark log
  + will auto update once, after that no more
- Even when the lifetime is over, `rdcli info` still says that it is registered
  Step to reproduce:
  + `rdcli reg [rd_ip]` and wait until lifetime over
  + `rdcli info` still says registered to RD
- Calling update after lifetime is over make program hang
  Step to reproduce:
  + `rdcli reg [rd_ip]` and wait until lifetime over. Note: the client auto updates once, so wait at least 2 * lifetime
  + `rdcli update` and program hangs
- Registering to new RD, while already registered to another, the old registration will be deleted even though the new registration failed. I could not find about this in the draft.

Feedback:
- When using discover/reg with random ip, it takes too long until the shell command returns. Maybe more verbose output like ping6 helps, so that the user does not think that the program hangs somewhere
- There is no command for lookup
- Output of `rdcli rem` when not registered can be misleading. It says `node successfully removed from RD` when there are actually no nodes removed.

haukepetersen commented:
@pokgak Thanks for your feedback, I will see what I can do to address the issues you found. I seem to have broken the update procedure somehow, will fix that.

> Registering to new RD, while already registered to another, the old registration will be deleted even though the new registration failed. I could not find about this in the draft.

This has, as I see it, nothing to do with the draft, but is implementation specific behavior. I would like to leave it as is for now, but maybe we can improve this in the future. For info: I actually plan to allow the lib to connect to more than a single server, but this needs some improvements in our CoAP library first. Once that change will be made, the local handling of registrations will implicitly also be improved and more clear.

> Feedback [...]

Keep in mind, that the shell commands are rather a debugging instrument and not so much meant for creating responsive user interfaces... So having some (blocking) delay in some functions is IMHO completely file. Also keeping status messages of some coarse level of granularity is fine (as it saves memory), so I don't like to have an output message for each possible error case. But I will see what I can do to improve the situation at least a little bit.

waehlisch commented:
One remark: The current naming of the PR and functionality is confusing. This PR is about the implementation of an RD endpoint (EP) and not an RD client. Based on the [draft](https://tools.ietf.org/html/draft-ietf-core-resource-directory-15#section-3.2), the EP implements the registration and management interface. An RD client implements the lookup and group interface.

haukepetersen commented:
Fixed (hopefully all) of the findings:

> Client updates automatically once before the lifetime is over, but not anymore after that.

fixed: bug was that the update and remove functions got stuck when called from a different thread (`_waiter` was not updated when calling these functions...)

> Even when the lifetime is over, rdcli info still says that it is registered

fixed. The `update` function does not get stuck anymore, so the internal state handling should now be correct

> Calling update after lifetime is over make program hang

also fixed, was triggered by the same bug as the first finding

> Registering to new RD, while already registered to another, the old registration will be deleted even though the new registration failed. I could not find about this in the draft.

See comment above

> When using discover/reg with random ip, it takes too long until the shell command returns. Maybe more verbose output like ping6 helps, so that the user does not think that the program hangs somewhere

More output would be too much (as said, we not aiming for a rich user experience with the shell...). But I added an additional message that it could take a short while.

> Output of rdcli rem when not registered can be misleading. It says node successfully removed from RD when there are actually no nodes removed.

Fixed, the return value of `rdcli_remove()` is now more to the point.

haukepetersen commented:
Ups, forgot to mention that I did some optimizations on the code as well:
- split the `rdcli.h` header but moving the `rdcli_standalone_x` stuff in a separate header
- added the possibility to add an event callback to `rdcli_standalone`. Now a user can simply react to certain registration events
- use that event callback in the example application to output info on state changes
- updated also some error messages in the shell commands
- renamed the `rdcli rem` command to `rdcli remove` to be more consistent with the other commands

waehlisch commented:
@haukepetersen you should at least mention in the doc (very upfront) that you implement an RD endpoint, which is named RD client in this implementation. Just to prevent confusion

PeterKietzmann commented:
@pokgak some changes have been made. Wanna test again?

waehlisch commented:
@pokgak can you please also test that the registration is idempotent, see https://tools.ietf.org/html/draft-ietf-core-resource-directory-15#section-5.3?

haukepetersen commented:
> @haukepetersen you should at least mention in the doc (very upfront) that you implement an RD endpoint, which is named RD client in this implementation. Just to prevent confusion

Yes, will do.

haukepetersen commented:
fixed two more bugs / debug leftovers that I missed...

waehlisch commented:
@pokgak 
> Registering to new RD, while already registered to another, the old registration will be deleted even though the new registration failed. I could not find about this in the draft.

Did you verify the behavior described on page 26: "If the registration fails with a Service Unavailable response and a Max-Age option or Retry-After header, the registering endpoint SHOULD retry the operation after the time indicated. If the registration fails in another way, including request timeouts, or if the Service Unavailable error persists after several retries, or indicates a longer time than the endpoint is willing to wait, it SHOULD pick another registration URI from the "URI Discovery" step and if there is only one or the list is exhausted, pick other choices from the "Finding a Resource Directory" step.  "

haukepetersen commented:
one more 'optimization' while I was at it: renamed `rdcli reg` to `rdcli register`. Now all sub-commands have the same style.

haukepetersen commented:
> Did you verify the behavior described on page 26:

No need to verify, as the described behavior is not implemented (yet). I skipped it in the implementation, as I think the benefit of fully adhering to it in our context of constrained nodes is not worth the additional implementation complexity and memory requirements.

Regarding the first part: implementing 'high-level' retries on the application layer is IMHO overkill and would require a quite different architecture for the `rdcli` library. I would actually see this behavior to be implement on top of what is proposed here. But we might want to optimize the return message of `rdcli_register` some point in the future, to reflect any `service unavailable + x` responses.

Regarding the second part: I see this as well to rather be handled by the application using the `rdcli` library...

chrysn commented:
@waehlisch wrote:
> pokgak can you please also test that the registration is idempotent

That'd be a property of the server; unless gcoap offers additional optimization for known-idempotent requests, there's nothing to be done for the registrant-ep.

waehlisch commented:
> 
> 
> @waehlisch wrote:
> 
> > pokgak can you please also test that the registration is idempotent
> 
> That'd be a property of the server; unless gcoap offers additional optimization for known-idempotent requests, there's nothing to be done for the registrant-ep.

Oops, completely right. My bad! Sorry.

chrysn commented:
> >  it SHOULD pick another registration URI from the "URI Discovery" step
>
> I skipped it in the implementation, as I think the benefit of fully adhering to it in our context of constrained nodes is not worth the additional implementation complexity and memory requirements.

There is the possibility of random picking -- that'd be a very controversial thing to say in the draft, but I think it still might be a good compromise between keeping track and not implementing anything of this. If you, at any choice you get (eg. if there's multiple links in the URI discovery step, if you found several RDs in the ND option or if a given hostname resolves to several AAAA records), pick a random one, then it's sufficient if you let the application decide when to retry after a failed registration, but you don't have to carry state around and still eventually wind up with the working server in a redundant setup.

pokgak commented:
Tested again, auto update now works. Discover, register, remove, info, update all OK.

Also, I'm curious as to why lookup is not implemented. Is it because it is basically the same as discovery, just with different query parameter?

haukepetersen commented:
> Also, I'm curious as to why lookup is not implemented. Is it because it is basically the same as discovery, just with different query parameter?

Yes, but it still takes additional time to do it, which I did not have. So please feel free to add this functionality in a follow up PR :-)

haukepetersen commented:
rebased and fixed a minor issue about a removed config value that is actually used by `rdcli_simple`

haukepetersen commented:
@pokgak @bergzand despite the naming issue (which I will address in a follow up PR), do you give your go so we can get this PR merged?!

pokgak commented:
ACK from my side =)

haukepetersen commented:
@pokgak very nice, thanks for the testing!

@bergzand I will squash now, hope thats ok with you...

bergzand commented:
> @bergzand I will squash now, hope thats ok with you...

Go ahead, I didn't have time to look at this yet unfortunately :(

haukepetersen commented:
blacklisted boards with insufficient memory

haukepetersen commented:
fixed a parenthesis issues that clang did not like...

haukepetersen commented:
more blacklisting...

jia200x commented:
as far as I see this PR has chances to be merged for the Release, right?

miri64 commented:
Codacy has a problem, but I don't see what it is (it just tells "1 warning" but not showing the actual warning). To my knowledge however, Murdock is still the canonical check (and also the only required one).

miri64 commented:
So go!

haukepetersen commented:
Finally :-) Thanks everyone!

kb2ma commented:
Nice, Hauke! I just updated the [Standards Based Application Development](https://github.com/RIOT-OS/RIOT/wiki/Standards-Based-Application-Development) wiki page so hopefully the implementation and apps will be easy to find.

waehlisch commented:
But we still should work on aligning naming with the RFC.

jia200x commented:
awesome! Thanks to everyone involved! :tada:

haukepetersen commented:
> But we still should work on aligning naming with the RFC.

done: #10163
